File: include/nol.py
import time

class Document:
    def __init__(self, root): self.root = root
    def get(self, path):
        curr = self.root
        for p in path.split('.'):
            if not isinstance(curr, dict) or p not in curr: return None
            curr = curr[p]
        return curr
    def exists(self, path): return self.get(path) is not None
    def to_dict(self): return self.root

class Builder:
    def __init__(self): self.root = {}
    def set(self, path, value):
        parts = path.split('.')
        curr = self.root
        for p in parts[:-1]:
            if p not in curr or not isinstance(curr[p], dict): curr[p] = {}
            curr = curr[p]
        curr[parts[-1]] = value
        return self
    def build(self): return Document(self.root)

class Parser:
    def __init__(self, text): self.text, self.pos, self.start, self.depth = text, 0, time.time(), 0
    def peek(self): return self.text[self.pos] if self.pos < len(self.text) else None
    def advance(self):
        if self.pos >= len(self.text): return None
        c = self.text[self.pos]; self.pos += 1; return c
    def skip_ws(self):
        while self.pos < len(self.text):
            if self.text[self.pos].isspace(): self.advance()
            elif self.text[self.pos] == '#':
                self.advance()
                if self.peek() == '#':
                    self.advance()
                    while self.pos < len(self.text) - 1 and self.text[self.pos:self.pos+2] != '##': self.advance()
                    if self.pos < len(self.text): self.advance(); self.advance()
                else:
                    while self.pos < len(self.text) and self.text[self.pos] != '\n': self.advance()
            else: break
    def read_str(self, q):
        s = ""
        while self.pos < len(self.text):
            c = self.advance()
            if c == q: return s
            if c == '\\':
                e = self.advance()
                if e == 'u': s += chr(int(self.text[self.pos:self.pos+4], 16)); self.pos += 4
                elif e == 'U': s += chr(int(self.text[self.pos:self.pos+8], 16)); self.pos += 8
                else: s += {'n':'\n', 'r':'\r', 't':'\t'}.get(e, e)
            else: s += c
        return s
    def read_key(self):
        self.skip_ws(); c = self.peek()
        if c in ('"', "'"): return self.read_str(self.advance())
        k = ""
        while self.pos < len(self.text) and (self.text[self.pos].isalnum() or self.text[self.pos] in '_-'): k += self.advance()
        return k
    def parse_value(self):
        self.skip_ws()
        if time.time() - self.start > 1: raise Exception("Timeout")
        self.depth += 1
        if self.depth > 100: raise Exception("Depth")
        c = self.peek()
        if c == '{':
            self.advance(); o = {}
            while self.peek() and self.peek() != '}':
                self.skip_ws(); k, v = self.parse_pair(); o[k] = v
                self.skip_ws()
                if self.peek() == ',': self.advance()
            if self.peek() == '}': self.advance()
            res = o
        elif c == '[':
            self.advance(); a = []
            while self.peek() and self.peek() != ']':
                self.skip_ws(); a.append(self.parse_value())
                self.skip_ws()
                if self.peek() == ',': self.advance()
            if self.peek() == ']': self.advance()
            res = a
        elif c in ('"', "'"): res = self.read_str(self.advance())
        elif c and (c.isdigit() or c == '-'):
            s = ""
            while self.pos < len(self.text) and (self.text[self.pos].isdigit() or self.text[self.pos] in '.-eE+'): s += self.advance()
            res = float(s) if '.' in s or 'e' in s.lower() else int(s)
        else:
            s = ""
            while self.pos < len(self.text) and self.text[self.pos].isalpha(): s += self.advance()
            if s == "true": res = True
            elif s == "false": res = False
            elif s == "null": res = None
            else: raise Exception(f"Invalid value: {s}")
        self.depth -= 1; return res
    def parse_pair(self):
        k = self.read_key()
        if k in ("_env", "_interpolate", "_meta"): raise Exception(f"Reserved: {k}")
        self.skip_ws()
        if self.peek() == ':': self.advance()
        return k, self.parse_value()
    def parse(self):
        root = {}
        while self.pos < len(self.text):
            self.skip_ws()
            if not self.peek(): break
            if self.peek() == '[':
                self.advance(); path = ""
                while self.peek() and self.peek() != ']': path += self.advance()
                if self.peek() == ']': self.advance()
                parts = path.split('.'); curr = root
                for p in parts[:-1]:
                    if p not in curr or not isinstance(curr[p], dict): curr[p] = {}
                    curr = curr[p]
                last = parts[-1]
                if last not in curr or not isinstance(curr[last], dict): curr[last] = {}
                self.parse_into(curr[last])
            else:
                k, v = self.parse_pair()
                if k in root: raise Exception(f"Dup: {k}")
                root[k] = v
        return Document(root)
    def parse_into(self, o):
        while self.pos < len(self.text):
            self.skip_ws()
            if self.peek() == '[' or not self.peek(): break
            k, v = self.parse_pair()
            if k in o: raise Exception(f"Dup: {k}")
            o[k] = v

def parse(text): return Parser(text).parse()

File: include/nol.js
class Document {
    constructor(root) { this.root = root; }
    get(path) {
        let curr = this.root;
        for (const p of path.split('.')) {
            if (curr === null || typeof curr !== 'object' || !(p in curr)) return null;
            curr = curr[p];
        }
        return curr;
    }
    exists(path) { return this.get(path) !== null; }
    toObject() { return this.root; }
}

class Builder {
    constructor() { this.root = {}; }
    set(path, value) {
        const parts = path.split('.');
        let curr = this.root;
        for (let i = 0; i < parts.length - 1; i++) {
            const p = parts[i];
            if (!(p in curr) || typeof curr[p] !== 'object') curr[p] = {};
            curr = curr[p];
        }
        curr[parts[parts.length - 1]] = value;
        return this;
    }
    build() { return new Document(this.root); }
}

class Parser {
    constructor(text) { this.text = text; this.pos = 0; this.start = Date.now(); this.depth = 0; }
    peek() { return this.text[this.pos] || null; }
    advance() { return this.text[this.pos++] || null; }
    skipWs() {
        while (this.pos < this.text.length) {
            if (/\s/.test(this.text[this.pos])) this.advance();
            else if (this.text[this.pos] === '#') {
                this.advance();
                if (this.peek() === '#') {
                    this.advance();
                    while (this.pos < this.text.length - 1 && this.text.slice(this.pos, this.pos + 2) !== '##') this.advance();
                    if (this.pos < this.text.length) { this.advance(); this.advance(); }
                } else {
                    while (this.pos < this.text.length && this.text[this.pos] !== '\n') this.advance();
                }
            } else break;
        }
    }
    readStr(q) {
        let s = "";
        while (this.pos < this.text.length) {
            const c = this.advance();
            if (c === q) return s;
            if (c === '\\') {
                const e = this.advance();
                if (e === 'u') { s += String.fromCharCode(parseInt(this.text.slice(this.pos, this.pos + 4), 16)); this.pos += 4; }
                else if (e === 'U') { s += String.fromCodePoint(parseInt(this.text.slice(this.pos, this.pos + 8), 16)); this.pos += 8; }
                else { s += { 'n': '\n', 'r': '\r', 't': '\t' }[e] || e; }
            } else s += c;
        }
        return s;
    }
    readKey() {
        this.skipWs(); const c = this.peek();
        if (c === '"' || c === "'") return this.readStr(this.advance());
        let k = "";
        while (this.pos < this.text.length && /[a-zA-Z0-9_-]/.test(this.text[this.pos])) k += this.advance();
        return k;
    }
    parseValue() {
        this.skipWs();
        if (Date.now() - this.start > 1000) throw new Error("Timeout");
        if (++this.depth > 100) throw new Error("Depth");
        const c = this.peek(); let res;
        if (c === '{') {
            this.advance(); const o = {};
            while (this.peek() && this.peek() !== '}') {
                this.skipWs(); const [k, v] = this.parsePair(); o[k] = v;
                this.skipWs(); if (this.peek() === ',') this.advance();
            }
            if (this.peek() === '}') this.advance();
            res = o;
        } else if (c === '[') {
            this.advance(); const a = [];
            while (this.peek() && this.peek() !== ']') {
                this.skipWs(); a.push(this.parseValue());
                this.skipWs(); if (this.peek() === ',') this.advance();
            }
            if (this.peek() === ']') this.advance();
            res = a;
        } else if (c === '"' || c === "'") res = this.readStr(this.advance());
        else if (c && (/[0-9]/.test(c) || c === '-')) {
            let s = "";
            while (this.pos < this.text.length && /[0-9.-eE+]/.test(this.text[this.pos])) s += this.advance();
            res = s.includes('.') || /e/i.test(s) ? parseFloat(s) : parseInt(s, 10);
        } else {
            let s = "";
            while (this.pos < this.text.length && /[a-zA-Z]/.test(this.text[this.pos])) s += this.advance();
            if (s === "true") res = true;
            else if (s === "false") res = false;
            else if (s === "null") res = null;
            else throw new Error("Invalid value: " + s);
        }
        this.depth--; return res;
    }
    parsePair() {
        const k = this.readKey();
        if (["_env", "_interpolate", "_meta"].includes(k)) throw new Error("Reserved: " + k);
        this.skipWs(); if (this.peek() === ':') this.advance();
        return [k, this.parseValue()];
    }
    parse() {
        const root = {};
        while (this.pos < this.text.length) {
            this.skipWs(); if (!this.peek()) break;
            if (this.peek() === '[') {
                this.advance(); let path = "";
                while (this.peek() && this.peek() !== ']') path += this.advance();
                if (this.peek() === ']') this.advance();
                const parts = path.split('.'); let curr = root;
                for (let i = 0; i < parts.length - 1; i++) {
                    const p = parts[i];
                    if (!(p in curr) || typeof curr[p] !== 'object') curr[p] = {};
                    curr = curr[p];
                }
                const last = parts[parts.length - 1];
                if (!(last in curr) || typeof curr[last] !== 'object') curr[last] = {};
                this.parseInto(curr[last]);
            } else {
                const [k, v] = this.parsePair();
                if (k in root) throw new Error("Dup: " + k);
                root[k] = v;
            }
        }
        return new Document(root);
    }
    parseInto(o) {
        while (this.pos < this.text.length) {
            this.skipWs(); if (this.peek() === '[' || !this.peek()) break;
            const [k, v] = this.parsePair();
            if (k in o) throw new Error("Dup: " + k);
            o[k] = v;
        }
    }
}

function parse(text) { return new Parser(text).parse(); }
module.exports = { parse, Document, Builder };
